### PR TITLE
Dataexchange filter delete

### PR DESCRIPTION
**Pull request by jwtlake**
## Overview
Added filter delete sub-controller and tests.

## Usage 
This route is used when the user deletes a filter from the request.

## Outcome
The sub controller takes the provided filter id and executes a DB function that deletes the filter.
**Returns:** If successful it returns an empty object.
```javascript 
{
  result: {
     _attributes: [],
  },
}
```
## DBFunction
**Name:** dx_delete_data_request_template_filter_f(filterID, Username)
**OnSuccess:** returns the filterId it just deleted 
**OnFailure:** throws a sql error

The function will error out if the request has already been submitted (so it's frozen) or if the user doesn't have access rights.

### CODE REVIEW

**Comment by swashcap** [comment]
What’s up with 32e669a6a0c50d5fd110df4bf4a10efeadc43833? The commit message doesn't make sense.

**Comment by swashcap** [comment]
Want to rebase 041f2fca6a30a903241d2c12c670246d8c7f25d2 onto the commit that made changes to _filter.js_?

File: packages/steelpenny/src/lib/controllers/dataexchange/filter.js [hunk 1]
```diff
@@ -109,7 +111,46 @@ const write = (bookshelf, query, cookie, groupId = null) => {
...
+ * @param  {Object} query  Uses the filter id provided in query.json._id
+ * @param  {Object} cookie
+ * @returns {Object} Completely meaningless object.  See returned value.
+ */
+const deleteById = (bookshelf, query, cookie) => {
+  const id = parseInt(query.json._id, 10);
+  const username = cookie.username;
+
+  const funcStr = 'SELECT dxdba.dx_delete_data_request_template_filter_f(?,?);';
+  const prams = [id, username];
```
Comment (swashcap) [inline packages/steelpenny/src/lib/controllers/dataexchange/filter.js]: Change the `prams` variable name to `params` for clarity.

File: packages/steelpenny/src/lib/controllers/dataexchange/filter.js [hunk 2]
```diff
@@ -109,7 +111,46 @@ const write = (bookshelf, query, cookie, groupId = null) => {
...
+
+  return bookshelf.knex.raw(funcStr, prams)
+  .then((response) => {
+    // verify that the function deleted the correct Filter_id
+    // TODO: This check might not be necessary
+    const responseId = parseInt(response.rows[0].dx_delete_data_request_template_filter_f, 10);
```
Comment (swashcap) [inline packages/steelpenny/src/lib/controllers/dataexchange/filter.js]: You could use [`lodash#get`](https://lodash.com/docs/4.17.1#get) to ensure prop looks don't explode 💣.

File: packages/steelpenny/src/lib/controllers/dataexchange/filter.js [hunk 3]
```diff
@@ -109,7 +111,46 @@ const write = (bookshelf, query, cookie, groupId = null) => {
...
+      throw boom.badImplementation(
+        'result error',
+        'Delete function returned a differnt Filter_id then what it was called with.'
+        + ` Called: ${id} Returned: ${responseId}`
+      );
+    }
+    // return empty object
+    return { _attributes: [] };
+  })
+  .catch((error) => {
```
Comment (swashcap) [inline packages/steelpenny/src/lib/controllers/dataexchange/filter.js]: Drop the `.catch` here. Let the root controller work it out.

File: packages/steelpenny/test/integration/dataexchange.js [hunk 4]
```diff
@@ -1302,369 +1349,398 @@ describe('Dataexchange tests', () => {
...
+
+        return apiClient.dataexchange.get(
+          'filter',
+          'delete',
+          JSON.stringify(queryJson),
+          cookie
+        )
+        .catch((error) => {
+          error.status.should.equal(400);
+          error.statusText.should.equal('Bad Request');
```
Comment (swashcap) [inline packages/steelpenny/test/integration/dataexchange.js]: Probably don't need to test this status text: it’s boom’s responsibility, really.

File: packages/steelpenny/test/unit/dataexchange.js [hunk 5]
```diff
@@ -4,10 +4,20 @@ const chai = require('chai');
...
+const filter = require('../../src/lib/controllers/dataexchange/filter');
 
 // Set should property of all objects for BDD assertions
 chai.should();
 
+// helper function for mocking up bookshelf responses
+const bookShelfOverride = (returnObj) => { // eslint-disable-line arrow-body-style
+  return {
+    knex: {
+      raw: () => new Promise((fulfill) => fulfill(returnObj)),
```
Comment (swashcap) [inline packages/steelpenny/test/unit/dataexchange.js]: Please use `Promise#resolve` here.

File: packages/steelpenny/test/unit/dataexchange.js [hunk 6]
```diff
@@ -56,7 +66,27 @@ describe('Dataexchange tests', () => {
...
+
+        const bookshelf = bookShelfOverride(returnObj);
+        const query = { json: { _id: '1001' } };
+        const cookie = { username: 'mochatest' };
 
+        return filter.delete(bookshelf, query, cookie)
+        .catch((error) => {
+          error.isBoom.should.equal(true);
+          error.output.statusCode.should.equal(500);
+          error.output.payload.error.should.equal('Internal Server Error');
```
Comment (swashcap) [inline packages/steelpenny/test/unit/dataexchange.js]: Drop the error text test, as per above.

File: packages/steelpenny/test/unit/dataexchange.js [hunk 7]
```diff
@@ -56,7 +66,27 @@ describe('Dataexchange tests', () => {
...
+          error.message.should.equal('result error');
+          error.data.should.equal(
+            'Delete function returned a differnt Filter_id then what it was called with.'
```
Comment (swashcap) [inline packages/steelpenny/test/unit/dataexchange.js]: Few changes here:


* Make this string one line
* `differnt` → `different`
* `then` → `than`

These error string changes should also be applied to the source _filter.js_ file.

File: packages/steelpenny/src/lib/controllers/dataexchange/filter.js [hunk 8]
```diff
@@ -109,7 +111,46 @@ const write = (bookshelf, query, cookie, groupId = null) => {
     });
 };
 
+/**
+ * Executes a DB function that deletes a filter 
+ *
+ * @param  {Object} bookshelf
```
Comment (rssk) [inline packages/steelpenny/src/lib/controllers/dataexchange/filter.js]: Flesh out docbloc function params w/ a bit more details

File: packages/steelpenny/test/integration/dataexchange.js [hunk 9]
```diff
@@ -1302,369 +1349,398 @@ describe('Dataexchange tests', () => {
...
 
-    it('Gets form attributes, with text option id', () => { // eslint-disable-line arrow-body-style
-      const queryJson = {
-        _parentId: null,
-        _attributes: [],
-        _type: 'Subjects',
-        _modalityId: 142,
-      };
+    describe('GET delete', () => {
+      it('deletes filters by id', () => {
```
Comment (drewlandis) [inline packages/steelpenny/test/integration/dataexchange.js]: line number may or may not be 713... find test by grep'ing for `should add another filter to the child group in the above request`

File: packages/steelpenny/src/lib/controllers/dataexchange/filter.js (hunk 1 above)
Comment (jwtlake) [inline packages/steelpenny/src/lib/controllers/dataexchange/filter.js]: updated

File: packages/steelpenny/src/lib/controllers/dataexchange/filter.js (hunk 2 above)
Comment (jwtlake) [inline packages/steelpenny/src/lib/controllers/dataexchange/filter.js]: implemented lodash.get

File: packages/steelpenny/test/integration/dataexchange.js (hunk 4 above)
Comment (jwtlake) [inline packages/steelpenny/test/integration/dataexchange.js]: removed

File: packages/steelpenny/test/unit/dataexchange.js (hunk 5 above)
Comment (jwtlake) [inline packages/steelpenny/test/unit/dataexchange.js]: updated

File: packages/steelpenny/test/unit/dataexchange.js (hunk 6 above)
Comment (jwtlake) [inline packages/steelpenny/test/unit/dataexchange.js]: removed

File: packages/steelpenny/lib/controllers/dataexchange/filter.js [hunk 10]
```diff
@@ -111,14 +111,16 @@ const write = (bookshelf, query, cookie, groupId = null) => {
...
+/* eslint-disable max-len */
 /**
- * Executes a DB function that deletes a filter 
+ * Executes a DB function that deletes a filter
  *
- * @param  {Object} bookshelf
- * @param  {Object} query  Uses the filter id provided in query.json._id
- * @param  {Object} cookie
- * @returns {Object} Completely meaningless object.  See returned value.
+ * @param  {Object}  bookshelf  Hapi ORM pluggin used to make DB calls
```
Comment (swashcap) [inline packages/steelpenny/lib/controllers/dataexchange/filter.js]: You don’t need the extra whitespace in the doc blocks: JSDoc can parse ‘em just fine. `s/ +/ /g`

File: packages/steelpenny/lib/controllers/dataexchange/filter.js [hunk 11]
```diff
@@ -111,14 +111,16 @@ const write = (bookshelf, query, cookie, groupId = null) => {
     });
 };
 
+/* eslint-disable max-len */
```
Comment (swashcap) [inline packages/steelpenny/lib/controllers/dataexchange/filter.js]: Same thing _dataexchange-group-delete_: don’t ignore maximum length on doc blocks.

File: packages/steelpenny/test/unit/dataexchange.js [hunk 12]
```diff
@@ -67,23 +67,17 @@ describe('Dataexchange tests', () => {
 
   describe('Filter controller', () => {
     describe('GET delete', () => {
-      it('Returns an error when the delete function returns a differnt id then expected', () => {
-        const returnObj = { rows: [{ dx_delete_data_request_template_filter_f: 999 }] };
-
-        const bookshelf = bookShelfOverride(returnObj);
+      it('Returns an error when the delete function returns a different id than expected', () => {
+        const bookshelf = bookShelfOverride({ rows: [{ dx_delete_data_request_template_filter_f: 999 }] }); // eslint-disable-line max-len
```
Comment (swashcap) [inline packages/steelpenny/test/unit/dataexchange.js]: Please don’t ignore the `max-len` rule if the line is code that can be easily broken up with new lines:

```diff
-        const bookshelf = bookShelfOverride({ rows: [{ dx_delete_data_request_template_filter_f: 999 }] });
+        const bookshelf = bookShelfOverride({
+          rows: [{
+            dx_delete_data_request_template_filter_f: 999,
+          }],
+        });
```

File: packages/steelpenny/test/unit/dataexchange.js [hunk 13]
```diff
@@ -67,23 +67,17 @@ describe('Dataexchange tests', () => {
 
   describe('Filter controller', () => {
     describe('GET delete', () => {
-      it('Returns an error when the delete function returns a differnt id then expected', () => {
-        const returnObj = { rows: [{ dx_delete_data_request_template_filter_f: 999 }] };
-
-        const bookshelf = bookShelfOverride(returnObj);
+      it('Returns an error when the delete function returns a different id than expected', () => {
```
Comment (swashcap) [inline packages/steelpenny/test/unit/dataexchange.js]: Your function doesn’t _return an error_, it _rejects_. Please update the description.

**Comment by swashcap** [comment]
@jwtlake This is good to go. Could you please resolve merge conflicts?

**Comment by jwtlake** [comment]
Fixed merge conflicts with a rebase on the latest dcrewrite.

**Comment by swashcap** [comment]
💯 🍡 🎆  Great work!

**Pull request merged**